applications/grandpa_verification.md

```diff
@@ -0,0 +1,580 @@
+
+We can split the plan into three main sections:
+
+- Preamble, which corresponds to sections 2 and 3 of the GRANDPA paper.
+- Safety, corresponds to section 4.1.
+- Liveness, corresponds to section 4.2.
+
+ Every section can also be divided into:
+
+- Search for required abstractions
+- Modification of proofs and discovery of 'technical lemmas'
+- Writing of the statements in the proof assistant without proofs
+- Proofs of the statements
+However, the '_Search of required abstractions_' would happen mainly in the '_Preamble section_'.
+A detailed list of the required steps for every section is presented below.
+
+### Preamble
+
+1. Sketch proofs of section 2 lemmas
+2. Split of section 2 proofs in more elemental lemmas
+3. Formal proofs of the lemmas and formal definition of the `g` function outside  the proof assistant
+4. Proofs refactorization in a way that can be followed in COQ code
+5. List all the required structures to define the `g` function and to express the GRANDPA algorithm
+6. Split the list of required structures in implementable ones and those with only interfaces
+7. Write the interface of all the structures in the list
+8. Write the statements of all the lemmas and the principal conclusions
+9. Implementation of the `g` function using only the interfaces
+10. Implementation of the GRANDPA protocol using only the interfaces
+11. Proofs of section 2 lemmas
+
+### Safety
+
+12. Refactor of proof sketches in section 4.1 using the developed abstractions.
+13. Split the proofs of section 4.1 in more elemental lemmas.
+14. Formal proofs of section 4.1 lemmas.
+15. Refine proofs to make them as close to COQ reasoning as possible.
+16. Write the statements to be proved in COQ.
+17. Proof of the lemma 4.2
+18. Proof of Theorem 4.1
+19. Proof of Corollary 4.3
+20. Proof of intermediate lemmas.
+
+### Liveness
+
+21. Refactor of proof sketches of section 4.2
+22. Split of proofs of section 4.2 in more elemental lemmas.
+23. Formal proofs of section 4.2 lemmas.
+24. Refine proofs to make them as close to COQ reasoning as possible.
+25. Write statements to be proved in COQ.
+26. Proof of lemma 4.4.
+27. Proof of lemma 4.5.
+28. Proof of lemma 4.6.
+29. Proof of lemma 4.7.
+30. Proof of lemma 4.8.
+31. Proof of lemma 4.9.
+32. Proof of intermediate lemmas.
+
+### Disclaimer on code extraction
+
+When we say code extraction, we don’t mean full implementation of the mentioned functions or data types. We refer to literal extraction of the functions/data types as is to `Haskell` code without extraction of the required definitions for the result to be a compilable program in `Haskell`.
+This allows us to replace some types and functions used in `COQ` with those already existing in the `Haskell` ecosystem.
+The complete extraction of the protocol including all the needed types would require a complete implementation of the protocol and is out of the scope of this proposal.
+
```

Comment:
Thanks for responding to my queries in detail. Could you please incorporate and consolidate those responses in the proposal itself. Maybe you can create a new subsection "Modelling Techniques".

Reply:
Hi, I have created a pull-request on this branch with the modeling and abstraction techniques we are planning to use on TLA+ once  @AltiMario approves it should show here.
Thank you.